### ISSUE TITLE
Scrambling does not solve anything

### ISSUE THREAD

**Issue by X-Ryl669** [noffle/web-udp#5]
Scrambling the content with a Xor key does not solve anything.

1. It prevents official usage of a browser for any existing UDP server
2. It does not provide any kind of security (false sense of security) because
2.1 One can send a bunch of zero to capture the Xor key used for scrambling (if mastering the client/browser code), or could keep track of zero in the protocol (where a zero is expected, the Xor's key byte will be disclaimed) if under external attack.
2.2 Then, since `(a ^ b) ^ c = a ^ (b ^ c)` and ` a ^ b ^ a = a`, one can bypass the scrambling by simply sending `expectedAttackMsg ^ key`.

It'd be much more simple to require the browser to grant permission to send UDP packets (like it does for WebRTC access's to camera and microphone). This would prevent any undesired attack.

**Comment by noffle** [comment]
Thanks for thinking this through @X-Ryl669.

How could the browser code capture the xor key? From the perspective of the code running on the website, it would call e.g. `WebUDP.send(ip, port, buffer)`. Even if it set an all-zero buffer for `buffer`, how would it learn what the final result sent out to `ip:port` was?

**Comment by X-Ryl669** [comment]
Either by a cooperating server that's echoing back received packet as-is via AJAX, or as a MITM HTTP proxy, or simply, by using any UDP server that's sending a known sequence, for example, DNS.

If it's doing `WebUDP.receive(dnsServer, dnsPort, buffer)` and can get the initial DNS reply via any other mean, the key is simply `key = buffer ^ expectedBuffer`.

In fact, it's either more easy, it just have to "attack" a DNS server (we don't care about the buffer), and fetch the reply. Since the DNS server will answer with an error (which is perfectly formatted and known), the key is found from `key = errorReplyFromDNSServerXored ^ expectedDNSErrorReply`.

In all cases, it's very weak. If you need to fiddle with UDP, let the user know, and require it allows it.

**Comment by noffle** [comment]
That makes sense: a fixed xor key shared between sending and receiving would be defeatable.

What if the xor key for sending is different from the one used for receiving? I was assuming this but I don't think I wrote it down anywhere. This would keep the browser code from ever learning the necesary information to craft packets capable of talking to existing non-web-udp endpoints.

**Comment by noffle** [comment]
Another idea is to have a loose handshake process between sides; something that only web-udp-capable endpoints could negotiate. This would prevent the browser code from having any ability to craft packets until after the endpoint was verified to be using web-udp as well.

I hear what you're saying about wanting to grant full open UDP permissions. I think there's a good argument for that as well, but this model would let users run something UDP-like without a permissions model (like websockets), which is what I'm more interested in.

**Comment by X-Ryl669** [comment]
The idea of xoring the data is not viable (even if you have 2 different keys) since you can still have a MITM that's computing the keys and transmitting them via a side channel.
A handshake might not work either for protocols when NAT punch-through is required (that is, most of UDP protocols of interest), since the other side can not "talk back". And honestly, web-socket are a pain to use, they don't support roaming, they close for whatever reasons. All in all, they are not *that* useful in the end, a simple Ajax solution works almost the same.

I would rather want a true, native UDP support with no constraint, than a pseudo SYN/ACK à la TCP on top of UDP, that would, de facto prevent any interoperability with existing protocols (like websocket). 

If only WebRTC wasn't such an elephant to implement, it has NAT passthrough, data channel and so on... I'd like to have a "simple" webrtc, where I can say, that byte at position X in this packet should be Y.

**Comment by noffle** [comment]
You could write a MITM with websockets too. I don't think that's preventable. Am I misunderstanding you?

Yeah agreed, handshake is not preferrable.

Overall, it sounds like we're interested in different things. Which is fine: if this project ever gets off the ground then you could mostly implement your pure UDP approach by just removing some code. :)

**Comment by X-Ryl669** [comment]
The main issue I'm seeing with UDP in browser is DDOS (unlike TCP, since there is no amplification factor because of ACK). Just make a website with some js that's targeting an IP address, and zam!, you've got a DDOS attacker network and users don't know about they are participating in this.

Scrambling the data, or handshaking does not solve this issue. In fact, it's even worse, since, from the attacked area, the incoming data is random.

The only mitigation I can think of, is to require the user's browser to ask for sending such packets.
It can be as convenient as "Remember my choice for this website", or "Remember my choice for this script source". User are now used to give permissions, so it should not be troublesome.

Also, it should allow any communication (no constraint on the protocol) to any destination.

**Comment by noffle** [comment]
You're totally right: scrambling doesn't prevent DDOSing -- only access to non-web-udp-aware services.

I was thinking that the user would be rate limited in some way until the server responded in a web-udp-compatible way. That would bound the outgoing traffic possible by a browser to unaware uncooperative UDP endpoints.

**Comment by hackergrrl** [comment]
Really appreciating your scrutiny! I've updated the readme with some clearer notes on these aspects.
